Allow user-generated arc line lists

tbowers7 opened the pull request:
This PR introduces the ability for users to specify their own line lists to add instrument-specific arc line identifications to a PypeIt reduction, and is primarily in response to #1352 .  This new feature is accomplished via:

- A new command-line script `pypeit_install_linelist` that places a user-specified line list file into the user's PypeIt cache.
- A new function `pypeit.data.get_linelist_filepath()` (paralleling other `get_*_filepath()` functions) that searches both the main code repo and the user's PypeIt cache for the specified file.

Updates to the documentation (`doc/wave_calib.rst`) are included to describe how to use the new functionality (including **big, bold warnings** about using NIST vacuum wavelengths).  Unit tests have been added to cover the new functions.

--------

In the preparation of this PR, several other minor changes were made (ironically making up the bulk of the files touched):

- Finessing the ugly hack in #1425 to clarify the formatting of `reid_arxiv` files.
- Persnickety documentation proofreading, including a reordering of topics in `doc/wave_calib.rst`.
- Moving unused functions from `pypeit/core/wavecal/waveio.py` to `pypeit/deprecated/waveio_old.py`, as discussed with @profxj .
- Adding a flag to `pypeit_identify` allowing for the residuals plot to auto-rescale in search of gross outliers to the fit (especially useful for manual echelle identifications).
- Fixing a nagging error in apparently unexecuted code in `pypeit/core/wavecal/autoid.py`, flagged by VSCode.
- Cleaning up older syntax in `pypeit/core/wavecal/waveio.py` as part of moving functions to `deprecated`.
- Conformal doc file changes.

tbowers7 committed:
Add ability to search the PypeIt cache for arc line lists

Infrastructure side of things to allow for users to call their own line lists
for wavelength calibration.  This commit adds a new `data.get_linelist_filepath()`
function, similar to other `get_*_filepath()` functions.  This case is slightly
different because the base linelists are still distributed with the ``pip``
package, but the functionality pulls user files from the cache, if they exist.

Next steps include building the install script for user line lists, and updating
the documentation to describe how to access this functionality.

	modified:   pypeit/core/wavecal/kdtree_generator.py
	modified:   pypeit/core/wavecal/templates.py
	modified:   pypeit/core/wavecal/waveio.py
	modified:   pypeit/data/arc_lines/convert_NIST_to_lists.py
	modified:   pypeit/data/utils.py
	new file:   pypeit/deprecated/waveio_old.py

tbowers7 committed:
Add more unit tests for `pypeit.data.utils`

	modified:   pypeit/tests/test_pkgdata.py

tbowers7 committed:
Add install script for user-created arc line lists

	modified:   pypeit/data/utils.py
	new file:   pypeit/scripts/install_linelist.py
	modified:   setup.cfg

tbowers7 committed:
Merge branch 'develop' into allow_user_linelist

tbowers7 committed:
Merge branch 'develop' into allow_user_linelist

tbowers7 committed:
Merge branch 'release' into allow_user_linelist

Include hotfix to `load_reid_arxiv()` with the goal of creating a more elegant
solution.

tbowers7 committed:
Merge branch 'develop' into allow_user_linelist

tbowers7 committed:
Adding a big TODO; add a "--version" option to `run_pypeit`

	modified:   pypeit/core/wavecal/waveio.py
	modified:   pypeit/scripts/run_pypeit.py

tbowers7 committed:
Merge branch 'develop' into allow_user_linelist

tbowers7 committed:
Merge branch 'telgrid_filepath_hotfix' into allow_user_linelist

tbowers7 committed:
Add option to `pypeit_identify` to allow residual plot to auto-rescale

Especially when performing new manual wavelength IDs for echelle spectrographs,
the residuals can become quite large.  By allowing the residual plot to auto-
rescale, outlier identifications can be more easily marked and dealt with.

This commit does not change the default behavior of `pypeit_identify`, and the
use of the new flag is reuqired to invoke the new functionality.

	modified:   pypeit/core/gui/identify.py
	modified:   pypeit/scripts/identify.py

tbowers7 committed:
Merge branch 'develop' into allow_user_linelist

tbowers7 committed:
Cleanup of changes ahead of PR submit

	modified:   pypeit/core/gui/identify.py
	modified:   pypeit/core/wavecal/autoid.py
	modified:   pypeit/scripts/install_linelist.py
	modified:   pypeit/scripts/run_pypeit.py

tbowers7 committed:
Finessing an ugly hack done for a `release` hotfix

	modified:   pypeit/core/wavecal/waveio.py
	modified:   pypeit/data/utils.py

tbowers7 committed:
Doc changes related to user-supplied arc line lists

Most of the changes are in `doc/wave_calib.rst`, but there are conformal and
typo changes elsewhere.

	modified:   doc/api/pypeit.rst
	modified:   doc/api/pypeit.scripts.rst
	modified:   doc/help/pypeit_identify.rst
	modified:   doc/wave_calib.rst
	modified:   pypeit/data/utils.py
	modified:   pypeit/scripts/__init__.py

tbowers7 committed:
Persnickety proofreading points

	modified:   wave_calib.rst

tbowers7 committed:
Fixing unit test to conform with return value

	modified:   pypeit/tests/test_pkgdata.py

tbowers7 commented:
Unit tests pass:
```
$ pytest -Wignore
============================= test session starts ==============================
platform linux -- Python 3.9.12, pytest-7.1.2, pluggy-1.0.0
...
collected 200 items                
...                                            
======================= 200 passed in 225.70s (0:03:45) ========================
```

This PR will require a cloud `dev-suite` run.

profxj commented on doc/help/pypeit_identify.rst:
```diff
@@ -31,7 +31,9 @@
       --pixtol PIXTOL       Pixel tolerance for Auto IDs (default: 0.1)
       --test                Unit tests? (default: False)
-      --linear              Show the spectrum in linear scale? (Default: log
+      --linear              Show the spectrum in linear (rather than log) scale?
```
nit:  I think we can now remove the `?`

profxj commented on doc/wave_calib.rst:
```diff
@@ -44,14 +48,122 @@ be certain by adding the following to the :doc:`pypeit_file`
+User-Supplied Line Lists
+------------------------
+
+Users may now employ their own line lists, especially those including
```
Somewhere, maybe here, we might add a line like:

`Users who are confident that their new lines would benefit PypeIt broadly, i.e. beyond this single instrument, should consult with the Developers.`

tbowers7 replied:
Added as a `..note`

profxj commented on pypeit/core/wavecal/waveio.py:
```diff
@@ -188,52 +151,12 @@ def load_line_list(line_file, add_path=False, use_ion=False, NIST=False):
-    return line_list
+    line_file = data.get_linelist_filepath(f'{line_file}_lines.dat') if use_ion else \
+        data.get_linelist_filepath(line_file)
+    return astropy.table.Table.read(line_file, format='ascii.fixed_width', comment='#')
```
whoa, that's so much cleaner!

profxj commented on pypeit/core/wavecal/waveio.py:
```diff
@@ -370,58 +266,6 @@ def load_tree(polygon=4, numsearch=20):
     return file_load, index
 
 
-def load_nist(ion):
```
boy, you are touching some of the oldest PypeIt code still around! :)

profxj commented on pypeit/tests/test_pkgdata.py:
```diff
@@ -18,32 +21,56 @@ def test_cloud_url():
+    assert isinstance(skyspec, xspectrum1d.XSpectrum1D)
+
+
+def test_search_cache():
```
thx for all these new tests!

profxj reviewed (approved):
Thx for this new functionality.  A couple of nits from
me but nothing important.

tbowers7 committed:
Addressing PR comments; catching additional doc files

	new file:   doc/api/pypeit.scripts.install_linelist.rst
	modified:   doc/help/pypeit_identify.rst
	new file:   doc/help/pypeit_install_linelist.rst
	new file:   doc/help/pypeit_ql_multislit.rst
	modified:   doc/wave_calib.rst
	modified:   pypeit/scripts/identify.py

profxj commented:
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  200 passed, 1195 warnings in 235.80s (0:03:55) ---
--- PYTEST UNIT TESTS PASSED  115 passed, 1803 warnings in 1165.96s (0:19:25) ---
--- PYTEST VET TESTS FAILED  1 failed, 23 passed, 1551 warnings in 934.18s (0:15:34) ---
--- PYPEIT DEVELOPMENT SUITE PASSED 152/152 TESTS  ---
Testing Started at 2022-07-26T20:24:30.037723
Testing Completed at 2022-07-27T07:11:12.119809
Total Time: 10:46:42.082086

profxj commented:
`test_identify` failed with this error:

spatid=str(slits.spat_id[slit]), pxtoler=pxtoler,
specname=specname, rescale_resid=rescale_resid)
E       UnboundLocalError: local variable 'wv_calib' referenced before assignment

/PypeIt/pypeit/core/gui/identify.py:351: UnboundLocalError
----------------------------- Captured stdout call -----------------------------

tbowers7 committed:
Responding to failed VET test

	modified:   CHANGES.rst
	modified:   doc/wave_calib.rst
	modified:   pypeit/core/gui/identify.py

tbowers7 committed:
Clarify some comments

	modified:   pypeit/core/gui/identify.py
	modified:   pypeit/scripts/identify.py

tbowers7 commented:
The `test_identify` VET test now passes:
```
$ pytest vet_tests/test_scripts.py::test_identify -Wignore
============================= test session starts ==============================
platform linux -- Python 3.9.12, pytest-7.1.2, pluggy-1.0.0
rootdir: /data/tbowers/PypeIt-development-suite
plugins: astropy-header-0.2.1, openfiles-0.5.0, remotedata-0.3.3, cov-3.0.0,
arraydiff-0.5.0, filter-subpackage-0.1.1, mock-3.7.0, doctestplus-0.12.0,
astropy-0.10.0, hypothesis-6.46.11
collected 1 item                                                               

vet_tests/test_scripts.py .                                              [100%]
============================== 1 passed in 5.54s ===============================
```

The fix was needing an `else` clause in `pypeit/core/gui/identify.py`, but this `else` clause has been missing for a long time.  Not sure why the VET test suddenly broke.  It seems related to these lines in `pypeit/scripts/identify.py`:
```
81        # Check if a solution exists
82        solnname = masterframe.construct_file_name(WaveCalib, msarc.master_key,
83                                                   master_dir=msarc.master_dir)
84        # wv_calib = waveio.load_wavelength_calibration(solnname) \
85        wv_calib = WaveCalib.from_file(solnname) \
85            if os.path.exists(solnname) and args.solution else None
```
where `wv_calib` is set to `None`.  This `None` is what is now caught in the new `else` clause added with commit 679784772065eb244ae87b4125ec6f9af6ee04b9.

tbowers7 committed:
Conformal doc change

	modified:   doc/help/pypeit_identify.rst

kbwestfall commented on doc/wave_calib.rst:
```diff
@@ -44,14 +48,131 @@ be certain by adding the following to the :doc:`pypeit_file`
+
+  Users who are confident that their new lines would benefit PypeIt
+  broadly (*i.e.*, beyond this single instrument) should consult with
+  the Developers.
```
Maybe we should just encourage them to submit a PR?  The "consulting with the Developers" can happen in the PR comments.

tbowers7 replied:
Adding encouragement.

kbwestfall commented on pypeit/core/wavecal/waveio.py:
```diff
@@ -441,74 +285,17 @@ def load_unknown_list(lines, unknwn_file=None, all=False):
-        tbl.write(f, format='ascii.fixed_width', overwrite=overwrite)
 
+    # Otherwise
+    msk = np.array([False]*len(line_list))
```
FWIW, you can also do this like so:
```
msk = np.zeros(len(line_list), dtype=bool)
```

tbowers7 replied:
Thanks!  I like the `np.zeros` version better.

kbwestfall reviewed (approved):
Wonderful!  I only had two very minor comments.  Approving now.

tbowers7 committed:
Final PR comments

	modified:   doc/wave_calib.rst
	modified:   pypeit/core/wavecal/waveio.py
